src/Microsoft.AspNet.Razor/CodeGenerators/CSharpTagHelperCodeRenderer.cs

```diff
@@ -438,27 +425,59 @@ private void RenderUnboundAttribute(string attributeName, Chunk attributeValueCh
-                    .WriteStartInstanceMethodInvocation(
-                        ExecutionContextVariableName,
-                        _tagHelperContext.ExecutionContextAddHtmlAttributeMethodName)
-                    .WriteStringLiteral(attributeName)
-                    .WriteParameterSeparator()
-                    .WriteStartMethodInvocation(_tagHelperContext.MarkAsHtmlEncodedMethodName);
+                string textValue = null;
+                var isPlainTextValue = TryGetPlainTextValue(attributeValueChunk, out textValue);
 
-                // If it's a plain text value then we need to surround the value with quotes.
                 if (isPlainTextValue)
                 {
-                    _writer.WriteStringLiteral(textValue);
+                    _writer
+                        .WriteStartInstanceMethodInvocation(
+                            ExecutionContextVariableName,
+                            _tagHelperContext.ExecutionContextAddHtmlAttributeMethodName)
+                        .WriteStringLiteral(attributeName)
+                        .WriteParameterSeparator()
+                        .WriteStartMethodInvocation(_tagHelperContext.MarkAsHtmlEncodedMethodName)
+                        .WriteStringLiteral(textValue)
+                        .WriteEndMethodInvocation(endLine: false)
+                        .WriteEndMethodInvocation();
+                }
+                else if (IsDynamicAttributeValue(attributeValueChunk))
+                {
+                    // Dynamic attribute value should be run through the conditional attribute removal system. It's
+                    // unbound and contains C#.
+
+                    _writer
+                        .WriteStartMethodInvocation(_tagHelperContext.AddHtmlAttributeValuesMethodName)
+                        .WriteStringLiteral(attributeName)
+                        .WriteParameterSeparator()
+                        .Write(ExecutionContextVariableName);
+
+                    _bodyVisitor.Accept(attributeValueChunk);
+
+                    _writer.WriteEndMethodInvocation();
                 }
                 else
                 {
+                    // HTML attributes are always strings. This attribute contains C# but is not dynamic. This occurs
+                    // when the attribute is a data-* attribute.
+
+                    // Attribute value is not plain text, mus be buffered to determine its final value.
+                    BuildBufferedWritingScope(attributeValueChunk, htmlEncodeValues: true);
+
+                    _writer
+                        .WriteStartInstanceMethodInvocation(
+                            ExecutionContextVariableName,
+                            _tagHelperContext.ExecutionContextAddHtmlAttributeMethodName)
+                        .WriteStringLiteral(attributeName)
+                        .WriteParameterSeparator()
+                        .WriteStartMethodInvocation(_tagHelperContext.MarkAsHtmlEncodedMethodName);
```

Comment:
Identical to lines 433-439. Suggest refactoring to shorten this method e.g. `RenderStartAddHtmlAttribute()`.